stratum/lib/security/credentials_manager.cc

```diff
@@ -13,26 +13,129 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-
 #include "stratum/lib/security/credentials_manager.h"
 
+#include <memory>
+#include <string>
+
+#include "absl/memory/memory.h"
 #include "gflags/gflags.h"
+#include "grpcpp/security/server_credentials.h"
+#include "grpcpp/security/server_credentials_impl.h"
+#include "grpcpp/security/tls_credentials_options.h"
 #include "stratum/glue/logging.h"
-#include "absl/memory/memory.h"
+#include "stratum/glue/status/status.h"
+#include "stratum/lib/macros.h"
+#include "stratum/lib/utils.h"
 
-namespace stratum {
+DEFINE_string(ca_cert, "", "CA certificate path");
+DEFINE_string(server_key, "", "gRPC Server pricate key path");
+DEFINE_string(server_cert, "", "gRPC Server certificate path");
 
-CredentialsManager::CredentialsManager() {}
+namespace stratum {
 
 CredentialsManager::~CredentialsManager() {}
 
+CredentialsManager::CredentialsManager() {
+  if (FLAGS_ca_cert.empty() || FLAGS_server_key.empty() ||
+      FLAGS_server_cert.empty()) {
+    LOG(WARNING) << "Using insecure server credentials";
+    server_credentials_ = ::grpc::InsecureServerCredentials();
+  } else {
+    // Load default credentials
+    ::util::Status status;
+    std::string pem_root_certs_;
+    std::string server_private_key_;
+    std::string server_cert_;
+    status.Update(ReadFileToString(FLAGS_ca_cert, &pem_root_certs_));
+    status.Update(ReadFileToString(FLAGS_server_key, &server_private_key_));
+    status.Update(ReadFileToString(FLAGS_server_cert, &server_cert_));
+    if (!status.ok()) {
+      LOG(WARNING)
+          << "Unable to load credentials, use insecure server credentials";
+      server_credentials_ = ::grpc::InsecureServerCredentials();
+      return;
+    }
+    auto credentials_reload_interface_ =
+        std::make_shared<CredentialsReloadInterface>(pem_root_certs_,
+                                                     server_private_key_,
+                                                     server_cert_);
+    auto credential_reload_config_ =
+        std::make_shared<TlsCredentialReloadConfig>(
+            credentials_reload_interface_);
+
+    tls_opts_ = std::make_shared<TlsCredentialsOptions>(
+        GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE,
+        GRPC_TLS_SERVER_VERIFICATION, nullptr,
+        credential_reload_config_, nullptr);
+    server_credentials_ = TlsServerCredentials(*tls_opts_);
+  }
+}
```

Comment:
I don't think we should call functions in constructors, that can fail: [Style Guilde](https://google.github.io/styleguide/cppguide.html#Doing_Work_in_Constructors)
[Abseil TotW #42](https://abseil.io/tips/42)

We often have these `CreateInstance` factory methods for that.

Reply:
Make sense, I follow the way Google did in AuthPolicyChecker, which adds an initialization function. It should be able to handle the error better.